File: pset6/credit/credit.py
from cs50 import get_string

# Credit card number validity checker


# Get credit card input
def get_credit_card():
    credit_card_number = get_string("What is your credit card number?\n")
    try:
        credit_card_number_int = int(credit_card_number)
        return credit_card_number
    except:
        return "INVALID"


# Check if Visa, MasterCard or Amex
def check_credit_card_company(first_2_chars, credit_card_len):
    # Check if Visa
    first_2_chars_int = int(first_2_chars)
    first_char_int = int(first_2_chars[0])
    second_char_int = int(first_2_chars[1])
    if (first_char_int == 4 and (credit_card_len == 13 or credit_card_len == 16)):
        return "VISA"
    elif ((first_2_chars_int == 34 or first_2_chars_int == 37) and credit_card_len == 15):
        return "AMEX"
    elif (first_char_int == 5 and (second_char_int > 0 and second_char_int < 6) and credit_card_len == 16):
        return "MASTERCARD"
    else:
        return "INVALID"


# Check if Luhn's Algorithm is met
def luhm_valid(cc_number_as_str):
    credit_card_len = len(cc_number_as_str)
    # Luhn 1 - Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together.
    i = -2
    luhn_1_product_digits = 0
    while (credit_card_len * -1) <= i:
        luhn_1_times_2 = int(cc_number_as_str[i]) * 2
        luhn_1_times_2_str = str(luhn_1_times_2)
        luhn_1_times_2_digits_summed = 0
        if (len(luhn_1_times_2_str) > 1):
            luhn_1_times_2_digits_summed = int(luhn_1_times_2_str[0]) + int(luhn_1_times_2_str[1])
        else:
            luhn_1_times_2_digits_summed = luhn_1_times_2
        luhn_1_product_digits = luhn_1_product_digits + luhn_1_times_2_digits_summed
        i = i - 2

    j = -1
    # Add the sum to the sum of the digits that weren’t multiplied by 2.
    luhn_2_sum_digits = luhn_1_product_digits
    while (credit_card_len * -1) <= j:
        luhn_2_sum_digits = luhn_2_sum_digits + int(cc_number_as_str[j])
        j = j - 2

    #If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
    if (luhn_2_sum_digits % 10 == 0):
        return check_credit_card_company(cc_number_as_str[:2], credit_card_len)
    else:
        return "INVALID"


credit_card = get_credit_card()
if(credit_card != "INVALID"):
    print(luhm_valid(credit_card))
else:
    print(credit_card)

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <math.h>

long get_credit_card_number(string prompt);

// Request user's credit card, determine if checksum is met and if yes, test to see if Amex, Visa or Mastercard
int main(void)
{

    long x = 0, last_digits, diff;

    int n, counter_luhn1 = 0, counter_luhn2 = 0, counter_luhn3 = 0, d1, d2, d3, d4 = 0;
    bool checksum;
    long credit_card = get_credit_card_number("What is your credit card number?\n");
    // Loop through all numbers in the provided credit card number
    for (int i = 1; i < 17; i++)
    {
        // Determine mod value to use and use it to isolate this iteration of loops final i number of digits in card number
        long to_mod_by_l = round(pow(10, i));
        last_digits = credit_card % to_mod_by_l;
        // Calculate difference between final i number of digits and previously identified numerator (from prior loop)
        diff = last_digits - x;
        // Determine digit at reverse i in the card number
        n = diff / (to_mod_by_l / 10);
        x = last_digits;
        // Capture first four digits in 16 digit string
        if (i > 12)
        {
            d4 = d3;
            d3 = d2;
            d2 = d1;
            d1 = n;
        }
        // Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together.
        if (i % 2 == 0)
        {
            int n2 = n * 2;
            int sum_of_prod_n;
            if (n2 / 10 >= 1)
            {
                sum_of_prod_n = 1 + n2 % 10;
            }
            else
            {
                sum_of_prod_n = n2 % 10;
            }
            counter_luhn1 = counter_luhn1 + sum_of_prod_n;
        }
        // Calculate the sum of the digits that weren’t multiplied by 2.
        else
        {
            counter_luhn2 = counter_luhn2 + n;
        }

    }

    // Sum two counter_luhn sets and check that mod 10 == 0 to confirm checksum indicates a valid credit card number
    counter_luhn3 = counter_luhn1 + counter_luhn2;
    if (counter_luhn3 % 10 == 0)
    {
        checksum = true;
    }
    else
    {
        checksum = false;
    }
    // Test to ensure length of credit card number is within range
    if (credit_card < 3999999999990 || credit_card > 9999999999999999)
    {
        printf("INVALID\n");
    }
    // Test to see if the length is 16 chars and first digits are 51, 52, 53, 54 or 55 - indicative of Mastercard
    else if (checksum == true && d1 == 5 && (d2 == 1 || d2 == 2 || d2 == 3 || d2 == 4 || d2 == 5))
    {
        printf("MASTERCARD\n");
    }
    // Test to see if the length is 15 chars and first digits are 34 or 37 - indicative of AMEX
    else if (checksum == true && d1 == 0 && d2 == 3 && (d3 == 4 || d3 == 7))
    {
        printf("AMEX\n");
    }
    // Test to see if the length is 13 or 16 chars and first digit is 4 - indicative of VISA
    else if (checksum == true && (d1 == 4 || (d1 == 0 && d2 == 0 && d3 == 0 && d4 == 4)))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

// Ensure user enters a credit card number
long get_credit_card_number(string prompt)
{
    long n;
    n = get_long("%s", prompt);
    return n;
}